<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Michael Sekula" />

<meta name="date" content="2020-06-23" />

<title>Getting started with the hbfm package</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Getting started with the hbfm package</h1>
<h4 class="author">Michael Sekula</h4>
<h4 class="date">2020-06-23</h4>



<p>This vignette provides an introduction to using the <code>hbfm</code> package for constructing a gene co-expression matrix from discrete single-cell RNA-sequencing (scRNA-seq) data. The <code>hbfm</code> package contains functions to implement the stochastic EM algorithm and MCMC sampler defined in the “A sparse Bayesian factor model for the construction of gene co-expression networks from single-cell RNA sequencing count data” manuscript.</p>
<p>To begin, load the <code>hbfm</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># Load the hbfm package</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(hbfm)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co"># Load example dataset from hbfm package</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="kw">data</span>(<span class="st">&quot;gene.mat&quot;</span>)</a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="kw">dim</span>(gene.mat)</a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="co">#&gt; [1]  10 100</span></a></code></pre></div>
<p>Functions in the <code>hbfm</code> package require discrete scRNA-seq gene expression data, so either a matrix or data frame of read counts is acceptable. The rows of the data correspond to genes and the columns correspond to cells. Here, the <code>gene.mat</code> dataset is a matrix of the expression counts of <em>G = 10</em> genes (rows) and <em>N = 100</em> cells (columns).</p>
<div id="stochastic-em" class="section level2">
<h2>Stochastic EM</h2>
<p>The first step in constructing a gene co-expression matrix with the <code>hbfm</code> package is to implement a stochastic EM algorithm with the <code>stoc.em</code> function to obtain a “good” set of initial starting values for the MCMC sampler. The <code>stoc.em</code> function requires two inputs:</p>
<ul>
<li><code>Y</code> - matrix or data.frame of gene expression count where the rows correspond to genes and columns correspond to cells; Y must contain intergets and have row names</li>
<li><code>Fac</code> - the number of latent factors to consider in the model</li>
</ul>
<p>The number of latent factors <code>Fac</code> is unknown, but should be less than the number of genes (<em>G</em>) in the dataset. Here, we will consider 5 factors in our example because our dataset only contains <em>G = 10</em> genes. However, for datasets with larger numbers of genes, we typically recommend choosing a value for <code>Fac</code> that is greater than 10 (e.g., 15, 20, or 25).</p>
<p>The other main parameters of the <code>stoc.em</code> function relate to the number of iterations in the stochastic EM algorithm. These parameters and their default values are listed below:</p>
<ul>
<li><code>M.stoc</code> - total number of stochastic EM iterations (default = 2000)</li>
<li><code>M.int</code> - initial number of MCMC draws before starting the EM algorithm (default = 100)</li>
<li><code>M.eval</code> - number iterations to be used for parameter estimation (default = 200)</li>
</ul>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">## Run stochastic EM first</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co">## Consider F=5 factors</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3">fit1 &lt;-<span class="st"> </span><span class="kw">stoc.em</span>(<span class="dt">Y=</span>gene.mat, <span class="dt">Fac =</span> <span class="dv">5</span>, <span class="dt">M.stoc =</span> <span class="dv">2000</span>, <span class="dt">M.int =</span> <span class="dv">100</span>, <span class="dt">M.eval =</span> <span class="dv">200</span>)</a></code></pre></div>
</div>
<div id="mcmc-sampler" class="section level2">
<h2>MCMC sampler</h2>
<p>The <code>hbfm.fit</code> function will run an MCMC sampler with the initial values generated by the <code>stoc.em</code> function. The only required input for <code>hbfm.fit</code> is <code>stoc.em.param</code>, which is the object created by the <code>stoc.em</code> function.</p>
<p>The other main parameters of the <code>hbfm.fit</code> function relate to the number of iterations in the MCMC algorithm. These parameters and their default values are listed below:</p>
<ul>
<li><code>M</code> - total number of MCMC iterations (default = 4000)</li>
<li><code>M.save</code> - number of iterations to calculate and save correlation estimates (default = 1000)</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">## Run MCMC sampler with initial parameter values from stoc.em</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">fit.res1 &lt;-<span class="st"> </span><span class="kw">hbfm.fit</span>(fit1, <span class="dt">M =</span> <span class="dv">4000</span>, <span class="dt">M.save =</span> <span class="dv">1000</span>)</a></code></pre></div>
</div>
<div id="gene-co-expression-matrix" class="section level2">
<h2>Gene co-expression matrix</h2>
<p>After running the MCMC sampler, the <code>corr.est</code> function is used to analyze the MCMC samples and obtain an estimated correlation matrix along with the upper and lower bounds of the 95% credible interval (CI). The input for the <code>corr.est</code> function is a list of objects created by the <code>hbfm.fit</code> function. In our example, we are inputting a list of just one object named <code>fit.res1</code>.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">## Obtain estimated gene-gene correlations from MCMC samples</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">fit.corr &lt;-<span class="st"> </span><span class="kw">corr.est</span>(<span class="kw">list</span>(fit.res1))</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="kw">print</span>(fit.corr)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co">#&gt;    gene.name1 gene.name2    corr  CI.low  CI.upp CI.sig    p.val</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co">#&gt; 1       Gene1      Gene2  0.3246  0.0960  0.6044    *   1.20e-02</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; 2       Gene1      Gene3 -0.1381 -0.5034  0.2798        5.46e-01</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#&gt; 3       Gene1      Gene4 -0.3717 -0.6584  0.0289        1.04e-01</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt; 4       Gene1      Gene5 -0.3310 -0.6499 -0.0281    *   2.00e-02</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt; 5       Gene1      Gene6 -0.9640 -1.0000 -0.8024    *   2.00e-05</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt; 6       Gene1      Gene7  0.2785 -0.1155  0.6104        1.52e-01</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt; 7       Gene1      Gene8  0.2776 -0.1080  0.6149        1.22e-01</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt; 8       Gene1      Gene9 -0.3488 -0.6498 -0.0816    *   1.20e-02</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">#&gt; 9       Gene1     Gene10  0.2260 -0.0111  0.4605        6.40e-02</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co">#&gt; 10      Gene2      Gene3 -0.5726 -1.0000  0.0099        6.00e-02</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt; 11      Gene2      Gene4  0.0827 -0.1991  0.2955        5.12e-01</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt; 12      Gene2      Gene5 -0.0750 -0.4461  0.2521        5.08e-01</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co">#&gt; 13      Gene2      Gene6 -0.2583 -0.5945  0.0000        1.20e-02</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="co">#&gt; 14      Gene2      Gene7  0.9400  0.7394  1.0000    *   2.00e-05</span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="co">#&gt; 15      Gene2      Gene8  0.0518 -0.3186  0.3314        6.16e-01</span></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="co">#&gt; 16      Gene2      Gene9  0.1292 -0.2026  0.3539        3.46e-01</span></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"><span class="co">#&gt; 17      Gene2     Gene10  0.2018 -0.2171  0.4620        2.48e-01</span></a>
<a class="sourceLine" id="cb4-22" data-line-number="22"><span class="co">#&gt; 18      Gene3      Gene4 -0.2060 -0.5630  0.2707        3.18e-01</span></a>
<a class="sourceLine" id="cb4-23" data-line-number="23"><span class="co">#&gt; 19      Gene3      Gene5  0.0666 -0.3508  0.4428        6.78e-01</span></a>
<a class="sourceLine" id="cb4-24" data-line-number="24"><span class="co">#&gt; 20      Gene3      Gene6  0.0868 -0.3741  0.4819        6.04e-01</span></a>
<a class="sourceLine" id="cb4-25" data-line-number="25"><span class="co">#&gt; 21      Gene3      Gene7 -0.4859 -1.0000  0.0418        7.00e-02</span></a>
<a class="sourceLine" id="cb4-26" data-line-number="26"><span class="co">#&gt; 22      Gene3      Gene8 -0.6526 -1.0000 -0.2335    *   1.20e-02</span></a>
<a class="sourceLine" id="cb4-27" data-line-number="27"><span class="co">#&gt; 23      Gene3      Gene9 -0.1533 -0.5639  0.2448        3.90e-01</span></a>
<a class="sourceLine" id="cb4-28" data-line-number="28"><span class="co">#&gt; 24      Gene3     Gene10 -0.1467 -0.5449  0.2735        5.10e-01</span></a>
<a class="sourceLine" id="cb4-29" data-line-number="29"><span class="co">#&gt; 25      Gene4      Gene5  0.0235 -0.2607  0.3168        8.44e-01</span></a>
<a class="sourceLine" id="cb4-30" data-line-number="30"><span class="co">#&gt; 26      Gene4      Gene6  0.4465 -0.0289  0.7141        1.04e-01</span></a>
<a class="sourceLine" id="cb4-31" data-line-number="31"><span class="co">#&gt; 27      Gene4      Gene7  0.0527 -0.2620  0.4155        6.90e-01</span></a>
<a class="sourceLine" id="cb4-32" data-line-number="32"><span class="co">#&gt; 28      Gene4      Gene8  0.0842 -0.2896  0.4863        6.70e-01</span></a>
<a class="sourceLine" id="cb4-33" data-line-number="33"><span class="co">#&gt; 29      Gene4      Gene9  0.9510  0.7529  1.0000    *   2.00e-05</span></a>
<a class="sourceLine" id="cb4-34" data-line-number="34"><span class="co">#&gt; 30      Gene4     Gene10 -0.0757 -0.3168  0.1232        5.22e-01</span></a>
<a class="sourceLine" id="cb4-35" data-line-number="35"><span class="co">#&gt; 31      Gene5      Gene6  0.3951  0.1309  0.6625    *   1.00e-02</span></a>
<a class="sourceLine" id="cb4-36" data-line-number="36"><span class="co">#&gt; 32      Gene5      Gene7 -0.0902 -0.4461  0.2372        4.02e-01</span></a>
<a class="sourceLine" id="cb4-37" data-line-number="37"><span class="co">#&gt; 33      Gene5      Gene8 -0.1710 -0.5219  0.2410        3.68e-01</span></a>
<a class="sourceLine" id="cb4-38" data-line-number="38"><span class="co">#&gt; 34      Gene5      Gene9  0.0408 -0.2615  0.3290        7.48e-01</span></a>
<a class="sourceLine" id="cb4-39" data-line-number="39"><span class="co">#&gt; 35      Gene5     Gene10 -0.8855 -1.0000 -0.6582    *   2.00e-05</span></a>
<a class="sourceLine" id="cb4-40" data-line-number="40"><span class="co">#&gt; 36      Gene6      Gene7 -0.2344 -0.5992  0.2357        1.82e-01</span></a>
<a class="sourceLine" id="cb4-41" data-line-number="41"><span class="co">#&gt; 37      Gene6      Gene8 -0.2406 -0.5845  0.1342        2.14e-01</span></a>
<a class="sourceLine" id="cb4-42" data-line-number="42"><span class="co">#&gt; 38      Gene6      Gene9  0.4229  0.1286  0.7004    *   4.00e-03</span></a>
<a class="sourceLine" id="cb4-43" data-line-number="43"><span class="co">#&gt; 39      Gene6     Gene10 -0.2918 -0.5973  0.0111        6.40e-02</span></a>
<a class="sourceLine" id="cb4-44" data-line-number="44"><span class="co">#&gt; 40      Gene7      Gene8 -0.0550 -0.6282  0.3083        8.97e-01</span></a>
<a class="sourceLine" id="cb4-45" data-line-number="45"><span class="co">#&gt; 41      Gene7      Gene9  0.1003 -0.2636  0.4851        4.94e-01</span></a>
<a class="sourceLine" id="cb4-46" data-line-number="46"><span class="co">#&gt; 42      Gene7     Gene10  0.2218 -0.0787  0.5008        1.68e-01</span></a>
<a class="sourceLine" id="cb4-47" data-line-number="47"><span class="co">#&gt; 43      Gene8      Gene9 -0.0218 -0.3467  0.3987        8.72e-01</span></a>
<a class="sourceLine" id="cb4-48" data-line-number="48"><span class="co">#&gt; 44      Gene8     Gene10  0.0608 -0.2682  0.3517        6.48e-01</span></a>
<a class="sourceLine" id="cb4-49" data-line-number="49"><span class="co">#&gt; 45      Gene9     Gene10 -0.0783 -0.3244  0.1334        4.82e-01</span></a></code></pre></div>
<p>The print output from the <code>corr.est</code> function provides a summary of all gene-gene pairs with the estimated correlation, the lower and upper bound of the 95% CI, and an approximate “p-value” that represents the proportion of posterior samples outside of the smallest CI that contains 0. The star in the CI.sig column denotes the gene pair correlations with CI’s that do not overlap with 0.</p>
<p>The gene co-expression matrix can also be obtained directly from the object created by <code>corr.est</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co">## Obtain gene co-expression matrix</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">fit.corr<span class="op">$</span>corr</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt;             Gene1       Gene2       Gene3       Gene4       Gene5</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; Gene1   1.0000000  0.32457581 -0.13811853 -0.37166705 -0.33099227</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="co">#&gt; Gene2   0.3245758  1.00000000 -0.57259075  0.08269666 -0.07496304</span></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co">#&gt; Gene3  -0.1381185 -0.57259075  1.00000000 -0.20595641  0.06658422</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="co">#&gt; Gene4  -0.3716670  0.08269666 -0.20595641  1.00000000  0.02348983</span></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co">#&gt; Gene5  -0.3309923 -0.07496304  0.06658422  0.02348983  1.00000000</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co">#&gt; Gene6  -0.9639709 -0.25834616  0.08684168  0.44648112  0.39509186</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10"><span class="co">#&gt; Gene7   0.2784775  0.93998072 -0.48592287  0.05272986 -0.09023091</span></a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="co">#&gt; Gene8   0.2775635  0.05175928 -0.65260649  0.08417365 -0.17100357</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co">#&gt; Gene9  -0.3487787  0.12915013 -0.15328175  0.95099483  0.04084802</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"><span class="co">#&gt; Gene10  0.2260246  0.20182132 -0.14667674 -0.07571672 -0.88551268</span></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co">#&gt;              Gene6       Gene7       Gene8       Gene9      Gene10</span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co">#&gt; Gene1  -0.96397094  0.27847750  0.27756345 -0.34877871  0.22602460</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="co">#&gt; Gene2  -0.25834616  0.93998072  0.05175928  0.12915013  0.20182132</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="co">#&gt; Gene3   0.08684168 -0.48592287 -0.65260649 -0.15328175 -0.14667674</span></a>
<a class="sourceLine" id="cb5-18" data-line-number="18"><span class="co">#&gt; Gene4   0.44648112  0.05272986  0.08417365  0.95099483 -0.07571672</span></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"><span class="co">#&gt; Gene5   0.39509186 -0.09023091 -0.17100357  0.04084802 -0.88551268</span></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co">#&gt; Gene6   1.00000000 -0.23443521 -0.24060077  0.42290877 -0.29176157</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21"><span class="co">#&gt; Gene7  -0.23443521  1.00000000 -0.05499286  0.10028675  0.22184682</span></a>
<a class="sourceLine" id="cb5-22" data-line-number="22"><span class="co">#&gt; Gene8  -0.24060077 -0.05499286  1.00000000 -0.02178234  0.06084387</span></a>
<a class="sourceLine" id="cb5-23" data-line-number="23"><span class="co">#&gt; Gene9   0.42290877  0.10028675 -0.02178234  1.00000000 -0.07826620</span></a>
<a class="sourceLine" id="cb5-24" data-line-number="24"><span class="co">#&gt; Gene10 -0.29176157  0.22184682  0.06084387 -0.07826620  1.00000000</span></a></code></pre></div>
</div>
<div id="model-selection" class="section level2">
<h2>Model selection</h2>
<p>Because the number of latent factors is unknown, model selection using the Deviance Information Criterion (DIC) can be used to find an appropriate number of factors. The input for the <code>hbfm.DIC</code> function is a list of objects created by the <code>hbfm.fit</code> function. In our example, we are inputting a list of just one object named <code>fit.res1</code>.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co">## Obtain DIC for model with 5 factors</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">hbfm.DIC</span>(<span class="kw">list</span>(fit.res1))</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#&gt; [1] 16756.42</span></a></code></pre></div>
</div>
<div id="extensions" class="section level2">
<h2>Extensions</h2>
<p>This vignette provides the details on how to use the different functions in the <code>hbfm</code> package. In a typical analysis, multiple chains would be run and different numbers of latent factors (<code>Fac</code>) would be considered. Both the <code>corr.est</code> function and <code>hbfm.DIC</code> function will combine the results from multiple runs of <code>hbfm.fit</code> if the input list consists of multiple objects created by this function. The choice of <code>Fac</code> with the lowest DIC would be considered the “best” model choice and the correlation estimates created with this model would be the most appropriate to use.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
